### ISSUE TITLE
Error when passing the JaxTuple as side-information of the loss function

### ISSUE THREAD

**Issue by zhongwen** [google/jax#483]
An issue related to #366, which is trying to pass the side-information out from grad.

The code to reproduce the error is pasted as follows.
```
from __future__ import print_function, division
import jax.numpy as np
from jax import grad, jit, vmap, vjp
from jax import random
from jax.tree_util import tree_map
from jax import core

from jax.experimental import optimizers

key = random.PRNGKey(0)

w = np.array([0., 1.])
x = random.normal(key, shape=(2,))
lr = np.array(1.)

opt_init, opt_update = optimizers.adam(1e-4)
opt_state = opt_init(w)

dw = grad(lambda w: np.square(np.dot(w, x) - 1.))
          
def loss(lr, opt_state):
  w = optimizers.get_params(opt_state)
  dummy_grad = dw(w) * lr
  opt_state = opt_update(0, dummy_grad, opt_state)
  w = optimizers.get_params(opt_state)
  return np.sum(np.square(w)), opt_state

# A trick from https://github.com/google/jax/issues/366
@jit
def mygrad(opt_state, lr):
  (l, extra), pullback = vjp(loss, lr, opt_state)
  all_zeros = tree_map(lambda x: core.pack(map(np.zeros_like, x)), opt_state)
  seeds = (np.asarray(1.0, np.float32), all_zeros)
  dx,_ = pullback(seeds)
  return dx, extra

grad, opt_state = mygrad(opt_state, lr)
```

Also the colab with the same code, if it's helpful https://colab.research.google.com/drive/1azubGbEk4ZS0EEVnm9iMyIwheDEN210m

**Comment by mattjj** [comment]
Try out the code in #484 if you can and let me know if that works for you!
